### ISSUE TITLE
docker run slurm but sinfo inval

### ISSUE THREAD

**Issue by Galen-Chen177** [nathan-hess/docker-slurm#18]
![image](https://user-images.githubusercontent.com/61819714/178399037-d10ced2a-2d89-4565-bf68-692045405d9d.png)
![image](https://user-images.githubusercontent.com/61819714/178399090-69dac6d1-b5d2-44a1-a58d-e26648e93cdd.png)
![image](https://user-images.githubusercontent.com/61819714/178399132-36e18ab8-193d-455f-a225-04875a92a8e4.png)
![image](https://user-images.githubusercontent.com/61819714/178399218-425f9d39-31b7-4a4d-9c83-228b34fceed4.png)

**Comment by Galen-Chen177** [comment]
I ran the simplest Slurm container, but he could not perform the task.

**Comment by nathan-hess** [comment]
Hi @Galen-Chen177!  Thanks for submitting this issue.

**Can you provide more detail on the setup you're using?**  Specifically, it looks like you're using Ubuntu, but can you share details on hardware, Linux distribution, whether you're running on a local or cloud server, etc.?

Just for future reference, if you can fill out the fields bug report template, it's enormously helpful in understanding the issue you're facing.  I'm more than happy to help you out with it, but I just need a little more information to understand what's going on.

---

I tried the same set of steps as you posted, and it seemed the container performed the expected behavior:
```Shell
C:\Users\nathan-hess>docker run --rm -d -it --hostname=linux --name=testcontainer nathanhess/slurm:full-root > NUL

C:\Users\nathan-hess>docker exec -it testcontainer bash
root@linux:~# squeue
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
root@linux:~# sinfo
PARTITION AVAIL  TIMELIMIT  NODES  STATE NODELIST
debug*       up   infinite      1   idle linux
root@linux:~# salloc -N 1 -n 1 -c 4
salloc: Granted job allocation 1
root@linux:~# srun echo $SLURM_JOBID
1
root@linux:~# exit
exit
salloc: Relinquishing job allocation 1
root@linux:~# exit
exit

C:\Users\nathan-hess>
```

**Comment by Galen-Chen177** [comment]
Thanks for reply, I deleted the previous iMage and re -executed a complete process. What information do you need? I don’t know Linux very much, sry
```
root@ip-10-0-1-205:/home/ubuntu# date
Tue Jul 12 03:47:31 UTC 2022
root@ip-10-0-1-205:/home/ubuntu# uname -a
Linux ip-10-0-1-205 5.4.0-1080-aws #87~18.04.1-Ubuntu SMP Fri Jun 10 18:32:41 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
root@ip-10-0-1-205:/home/ubuntu# docker pull nathanhess/slurm:full-root
full-root: Pulling from nathanhess/slurm
405f018f9d1d: Already exists 
d690a4ebea37: Pull complete 
b5a30c692f28: Pull complete 
128d6727418a: Pull complete 
efd7dd8f048d: Pull complete 
c59d90638ad4: Pull complete 
fdc92b43d1ab: Pull complete 
6605ca074d41: Pull complete 
4f4fb700ef54: Pull complete 
a29d4978f85b: Pull complete 
Digest: sha256:af77768626fe06234206f0ccb2d8999ecb8e101255b8888b850f0eaae33e5798
Status: Downloaded newer image for nathanhess/slurm:full-root
docker.io/nathanhess/slurm:full-root
root@ip-10-0-1-205:/home/ubuntu# docker run --rm -d -it --name=slurm nathanhess/slurm:full-root
e98e77316ad63716e8837d6ad625e5c511a6508d626725a9b29940c321aad5f2
root@ip-10-0-1-205:/home/ubuntu# docker exec -it slurm bash
root@e98e77316ad6:~# sinfo
PARTITION AVAIL  TIMELIMIT  NODES  STATE NODELIST
debug*       up   infinite      1  inval e98e77316ad6
root@e98e77316ad6:~# exit
exit
root@ip-10-0-1-205:/home/ubuntu# 
```

**Comment by nathan-hess** [comment]
Thanks for this information.  So it looks like you're running the container in an AWS instance?  Unfortunately I do not currently have access to AWS to test this out.  I tried running the same set of commands on Google Cloud, and it didn't give any issues.

The [INVAL node state code](https://slurm.schedmd.com/sinfo.html#OPT_INVAL) indicates that there's an issue registering the node with the Slurm controller.  One of the challenges about the setup in this image is that Slurm needs to know how many cores and how much memory to assign to the "compute node," but this can differ on every machine.  As a workaround, there's a script ([`/etc/startup.sh`](https://github.com/nathan-hess/docker-slurm/blob/cb20b0f95943827da8015fc5c765496e44c85b36/dockerfile_base/startup.sh#L14)) that runs the first time the container is started, and it edits `/etc/slurm/slurm.conf` to set these resources appropriately based on your system.  It's possible that this isn't working as expected, and perhaps the resources set in `slurm.conf` don't match the properties of your system.

A few things worth trying to diagnose this:
- Can you try running `cat /etc/slurm/slurm.conf` inside the running container?
- Can you try running `/etc/startup.sh` inside the running container?  When it runs, does it tell you that `slurmd` is already running?
- Are you able to run commands such as [`srun`](https://slurm.schedmd.com/srun.html) and [`sbatch`](https://slurm.schedmd.com/sbatch.html)?

A couple of questions:
- Are you able to say what context you're trying to use this image for?  Do you want to experiment with Slurm, trying to run automated tests, or something else?  I realize depending on the application you may not be able to answer this one in detail, no worries.
- How many cores and how much RAM does the VM you're running in have?  You should be able to find this by running `cat /proc/cpuinfo` (for cores) and `cat /proc/meminfo` and/or `free -h` (for memory).

**Comment by Galen-Chen177** [comment]
Yes, this is an instance of AWS (4 Cores, 16G MEMOMARY). There is an internal program in the company that requires SLURM as a scheduler. Because there is no SLURM environment, Debug is difficult, so I think of Docker
```
cat /etc/slurm/slurm.conf
...
# COMPUTE NODES
NodeName=e98e77316ad6 CPUs=4 RealMemory=15868  State=UNKNOWN
PartitionName=debug Nodes=ALL Default=YES MaxTime=INFINITE State=UP
```
```
root@e98e77316ad6:/etc# ./startup.sh 
 * Starting MUNGE munged                                                         [ OK ] 
 * Starting slurm compute node daemon slurmd                                     [ OK ] 
/usr/sbin/slurmd already running.
 * Starting slurm central management daemon slurmctld                            [ OK ] 
```
```
root@e98e77316ad6:/etc# srun hostname
srun: Required node not available (down, drained or reserved)
srun: job 1 queued and waiting for resources
^Csrun: Job allocation 1 has been revoked
srun: Force Terminated job 1
root@e98e77316ad6:/etc# sbatch --wrap='hostname'
Submitted batch job 2
root@e98e77316ad6:/etc# squeue
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
                 2     debug     wrap     root PD       0:00      1 (Nodes required for job are DOWN, DRAINED or reserved for jobs in higher priority partitions)
```
```
root@e98e77316ad6:/etc# free -h
               total        used        free      shared  buff/cache   available
Mem:            15Gi       8.4Gi       3.9Gi        59Mi       3.3Gi       6.8Gi
Swap:             0B          0B          0B
root@e98e77316ad6:/etc# cat /proc/cpuinfo |grep processor
processor       : 0
processor       : 1
processor       : 2
processor       : 3
```
`/etc/cpuinfo` and `/etc/slurm/slurm.conf` is too long .So I cut the section inside,do you need more infomation

**Comment by nathan-hess** [comment]
Wow, I'm quite surprised.  I've tested these images fairly extensively on Windows 10, even 21H1, and haven't had issues.  What version of Docker Desktop are you using?  And I'm assuming you're using the WSL 2 backend?

One option that _might_ provide a quick fix if you need it on AWS is to install the [APT package](http://packages.ubuntu.com/search?keywords=slurm-wlm).  It will provide a fairly basic Slurm installation, but that may be all you need for your use case.  Here's the [page on Slurm's website](https://slurm.schedmd.com/download.html) that mentions it.

I'm glad that it's functioning as expected on your work computer, but perplexed as to why the other Windows 10 computers aren't working as expected.  I'll investigate further and see if I can find any explanation or fixes.

**Comment by nathan-hess** [comment]
Sorry for the delay on this.  I think I've sorted out the issue.

### Root Cause

The issue appears to be caused by this line:

https://github.com/nathan-hess/docker-slurm/blob/cb20b0f95943827da8015fc5c765496e44c85b36/dockerfile_base/startup.sh#L15

When the container first starts, this line modifies this line of `/etc/slurm/slurm.conf`:

https://github.com/nathan-hess/docker-slurm/blob/cb20b0f95943827da8015fc5c765496e44c85b36/dockerfile_base/slurm.conf#L49

The reason this is done in a script is that the image can be used on any system, with any hostname, number of cores, and memory.  To avoid requiring the user to manually specify these details, this script automatically identifies and replaces these values.

However, the script obtains the memory from `/proc/meminfo`, where memory is specified in kibibits, so it must convert it to megabytes for `slurm.conf`.  In some cases, this calculation was rounded up, resulting in `slurm.conf` containing a value of `RealMemory` higher than the available system memory (as shown by `slurmd -C`).

This explains why the issue doesn't show up on every system -- it depends on the amount of memory on the system and how the calculated memory available to Slurm is rounded.


### Example

For instance, consider a Google Cloud Shell virtual machine.  The system memory is:

```
$ docker run --rm -it --hostname=linux nathanhess/slurm:full-root
root@linux:~# head -n 1 /proc/meminfo 
MemTotal:        8145584 kB
root@linux:~# slurmd -C               
NodeName=linux CPUs=2 Boards=1 SocketsPerBoard=1 CoresPerSocket=1 ThreadsPerCore=2 RealMemory=7954
```

However, the script sets `RealMemory` to $8145584/1024 \approx 7954.67 \approx 7955$ MB:

```
root@linux:~# tail -n 2 /etc/slurm/slurm.conf 
NodeName=linux CPUs=2 RealMemory=7955  State=UNKNOWN
PartitionName=debug Nodes=ALL Default=YES MaxTime=INFINITE State=UP
```

Thus, `RealMemory` gets set to a value (7955 MB) greater than the recognized system memory (7954 MB), resulting in the following error in the Slurm logs:

```
root@linux:~# cat /var/log/slurmctld.log
[...] error: Setting node linux state to DRAIN with reason:Low RealMemory (reported:7954 < 100.00% of configured:7955)
[...] drain_nodes: node linux state set to DRAIN
```

### Solution

There are a number of potential solutions, including:
- Set `RealMemory` by directly parsing the output of `slurmd -C`
- Decrease calculated memory by a small amount (perhaps 5%) so that it falls below the true system memory
  - However, this has the disadvantage that depending on the system, this may reduce available memory for Slurm by a non-negligible amount
- Require the end user to manually specify the number of cores and memory

The first option seems the best choice, as it requires no action from the user and since the memory is obtained from Slurm, it should always fall in an acceptable range.  This fix was implemented in #19.

### References
- StackOverflow or other forum posts
  - https://stackoverflow.com/questions/49902372/how-to-set-realmemory-in-slurm
  - https://stackoverflow.com/questions/71648940/how-is-realmemory-slurmd-c-calculated-why-does-it-differ-from-memtotal
- Slurm documentation
  - https://slurm.schedmd.com/slurmd.html
  - https://slurm.schedmd.com/slurm.conf.html

**Comment by nathan-hess** [comment]
Closing this issue since I _think_ it's fixed, but feel free to re-open or create a new issue if it still doesn't work as expected.
